docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,189 @@
+
+## Support for LDAP Group Schemas
+The two main LDAP schemas used to denote groups are 1) as first-class LDAP entries and 2) as LDAP attributes on LDAP first-class entries. In the first schema, members of the LDAP group would be listed in the first-class entry as attributes. In the second schema, LDAP first-class entries for users would list the groups that they are a member of in an attribute. These groups may or may not exist as first-class entries themselves.
+
+## Sync Process
+The `LDAPMemberLister` inside of the `LDAPGroupSyncer` will, for each group that needs to be synced, find the list of entries in the LDAP server that encode the members of the group. The `LDAPCachedMemberLister` has the same functionality for LDAP sync jobs where it is not possible to run an LDAP query to get the members of each group to be synced: when syncing all groups for an LDAP schema where groups are attributes, *etc*. The `LDAPMemberIdentityFactory` creates OpenShift `Identities` for these members. The `UserIdentityMapper` determines the OpenShift `Users` that these `Identities` refer to. Syncing all or some of either LDAP or OpenShift groups is supported.
+
+```go
+type LDAPGroupSyncer struct {
+  options            LDAPSyncOptions
+  filterGenerator    LDAPFilterGenerator
+  memberLister       LDAPMemberLister
+  cachedMemberLister LDAPCachedMemberLister
+  identityFactory    LDAPMemberIdentityFactory
+  identityMapper     *userapi.UserIdentityMapper
+  client             *osclient.Client
+}
+
+func (s *LDAPGroupSyncer) SyncAllInternalGroups() (err error) {
+  // use s.client to determine Groups to sync
+  s.SyncInternalGroups(groups)
+}
+
+func (s *LDAPGroupSyncer) SyncInternalGroups(groups *[]Group) (err error) {
+  // use s.filterGenerator to map Group -> groupFilter
+  s.SyncGroups(groupFilters)
+}
+
+func (s *LDAPGroupSyncer) SyncAllGroups() (err error) {
+  // use s.cachedMemberLister to determine groups to sync and find members for each
+  // for each group:
+  //    users := s.determineUsers(members)
+  // use s.groupClient to determine which Group to update
+}
+
+func (s *LDAPGroupSyncer) SyncGroups(groupFilters util.StringSet) (err error) {
+  // use s.memberLister to map groupFilter -> members
+  // for each group:
+  //    users := s.determineUsers(members)
+  // use s.groupClient to determine which Group to update
+}
+
+func (s *LDAPGroupSyncer) determineUsers(members []*ldap.Entry) (users []*userapi.User, err error) {
+  // use s.identityFactory to map member -> Identity
+  // use s.identityMapper to determine Identity -> user
+}
+
+```
+`LDAPSyncOptions` needs the following schema (TBD) :
+```
+LDAP server information (would this be in the identityprovider?)
+LDAP schema has first class groups?
+common LDAP filter for "all" (groups, users)
+specific filter key (for groups and for users)
+mapping OpenShift Group to specific filter value (either by opaque values or programatically from Group fields)
```

Comment:
The label only says what the provider of the data is, it does not contain the specific filter that, when conjoined with the common filter, retrieves the data required.